controllers/perconaservermysql_controller.go

```diff
@@ -733,10 +741,203 @@ func reconcileReplicationSemiSync(
+					}
+
+					l.Error(err, "Failed to run backup")
+					return
+				}
+
+				jobName := sch.Name + "-" + time.Now().Format("20060102150405")
+				l.Info("Running backup job",
+					"backup", jobName, "schedule", sch.Schedule, "storage", sch.StorageName)
+
+				bcp := &apiv1alpha1.PerconaServerMySQLBackup{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      jobName,
+						Namespace: cr.Namespace,
+						Labels: util.SSMapMerge(cr.Labels(), map[string]string{
+							"type":     "cron",
+							"ancestor": sch.Name,
+						}),
+					},
+					Spec: apiv1alpha1.PerconaServerMySQLBackupSpec{
+						ClusterName: cr.Name,
+						StorageName: sch.StorageName,
+					},
+				}
+
+				if err := r.Client.Create(context.TODO(), bcp); err != nil {
+					l.Error(err, "Failed to run backup")
+				}
+			})
+			if err != nil {
+				l.Error(err, "failed to add func to CronRegistry")
+				continue
+			}
+
+			r.CronRegistry.BackupJobs.Store(sch.Name, schedule.BackupJob{
+				BackupScheduleSpec: sch,
+				JobID:              jobID,
+			})
+		}
+	}
+
+	return nil
+}
+
+func (r *PerconaServerMySQLReconciler) deleteBackupJob(name string) {
+	job, ok := r.CronRegistry.BackupJobs.LoadAndDelete(name)
+	if !ok {
+		return
+	}
+	r.CronRegistry.Crons.Remove(job.(schedule.BackupJob).JobID)
+}
+
+type minHeap []apiv1alpha1.PerconaServerMySQLBackup
+
+func (h minHeap) Len() int { return len(h) }
+func (h minHeap) Less(i, j int) bool {
+	return h[i].CreationTimestamp.Before(&h[j].CreationTimestamp)
+}
+func (h minHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
+
+func (h *minHeap) Push(x interface{}) {
+	*h = append(*h, x.(apiv1alpha1.PerconaServerMySQLBackup))
+}
+func (h *minHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+func (r *PerconaServerMySQLReconciler) oldScheduledBackups(
 	ctx context.Context,
 	cr *apiv1alpha1.PerconaServerMySQL,
-) error {
+	ancestor string,
+	keep int,
+) ([]apiv1alpha1.PerconaServerMySQLBackup, error) {
+	backupList := apiv1alpha1.PerconaServerMySQLBackupList{}
+	err := r.Client.List(ctx, &backupList, &client.ListOptions{
+		Namespace: cr.Namespace,
+		LabelSelector: labels.SelectorFromSet(
+			util.SSMapMerge(cr.Labels(), map[string]string{
+				"type":     "cron",
+				"ancestor": ancestor,
+			}),
+		),
+	})
+	if err != nil {
+		return []apiv1alpha1.PerconaServerMySQLBackup{}, errors.Wrap(err, "list backups")
+	}
+
+	// nothing to delete
+	if len(backupList.Items) <= keep {
+		return []apiv1alpha1.PerconaServerMySQLBackup{}, nil
+	}
+
+	h := &minHeap{}
+	heap.Init(h)
+	for _, bcp := range backupList.Items {
+		if bcp.Status.State == apiv1alpha1.BackupSucceeded {
+			heap.Push(h, bcp)
+		}
+	}
+
+	if h.Len() <= keep {
+		return []apiv1alpha1.PerconaServerMySQLBackup{}, nil
+	}
+
+	old := make([]apiv1alpha1.PerconaServerMySQLBackup, 0, h.Len()-keep)
+	for i := h.Len() - keep; i > 0; i-- {
+		o := heap.Pop(h).(apiv1alpha1.PerconaServerMySQLBackup)
+		old = append(old, o)
+	}
```

Comment:
```suggestion
	succeeded := make([]apiv1alpha1.PerconaServerMySQLBackup, 0, len(backupList.Items))
	for _, bcp := range backupList.Items {
		if bcp.Status.State == apiv1alpha1.BackupSucceeded {
			succeeded = append(succeeded, bcp)
		}
	}
	if len(succeeded) <= keep {
		return []apiv1alpha1.PerconaServerMySQLBackup{}, nil
	}
	
	sort.Slice(succeeded, func(i, j int) bool {
		return succeeded[i].CreationTimestamp.Before(&succeeded[j].CreationTimestamp)
	})

	old := succeeded[0:len(succeeded)-keep]
```
Is there any need to use heap?